### PR TITLE
feat: Redis Pub/Sub Cache Invalidation (Story 3.3)

### PR DESCRIPTION
**Pull request by jamescrowley321**
## Summary
- Add `CacheInvalidationPublisher` with fire-and-forget Redis pub/sub on `identity:changes` channel
- Publish entity change events (entity_type, entity_id, operation, tenant_id, timestamp) after every canonical write across all 7 domain services (User, Role, Permission, Tenant, InboundSync, Reconciliation)
- Redis unavailability silently degrades — canonical writes are never affected (AC-3.3.2)
- Singleton publisher initialized in FastAPI lifespan with graceful degradation and URL scheme validation
- 29 new unit tests covering publisher, singleton lifecycle, and per-service integration

## Story
Refs #151
Part of PRD 5: Canonical Identity Domain Model

## Chained PR
Based on #211 — merge first

## Review Findings Addressed
- Security: 0 BLOCK, 2 WARN (W6 URL scheme validation, W7 singleton sentinel) — all resolved
- Blind Hunter: 4 MUST FIX, 7 SHOULD FIX, 4 NITPICK — all resolved
- Edge Cases: 3 paths — all resolved
- Acceptance: 2 PASS, 1 PARTIAL (publisher tests added for all services) — all resolved
- Red Team: no findings

## Test plan
- [x] Unit tests pass (`make test-unit`) — 1323 tests
- [x] Lint passes (`make lint`)
- [x] Independent review agents passed (2 iterations)
- [ ] CI passes

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by jamescrowley321** [comment]
## Review: Blind Hunter

### MUST FIX

- [`backend/app/services/inbound_sync.py:329`] Wrong operation label on create/link publish — uses `operation="sync"` for the create-or-link branch, but the correct operation for a newly-created user should be `"create"` (which IS used at line 319 for the existing-user update path). The variable `created` is available and distinguishes the two cases; consumers relying on `"create"` will miss cache invalidation for newly created users flowing through this path.

- [`backend/app/services/inbound_sync.py:349`] Wrong operation label on user deletion — `user.deleted` webhook deactivates a user but publishes `operation="sync"` instead of `operation="deactivate"`. The schema documents `"deactivate"` as a valid operation. Any consumer keyed on `"deactivate"` will never receive this event and will not invalidate caches for deactivated users.

- [`backend/uv.lock` (~line 1636, redis package entry)`] `redis` lock entry lists `pyjwt` as a dependency — `redis` has no dependency on `pyjwt`. This is a corrupt lockfile entry, likely a hand-edit or merge artifact. In production it is only harmless if `pyjwt` remains in the dependency tree via another package. If `pyjwt` is removed, `uv` would fail to install the environment. The lockfile must be regenerated with `uv lock` to produce a valid, reproducible entry.

- [`backend/app/main.py:95-103`] Leaked Redis connection when ping fails after TCP connect — on ping failure the `except Exception` block sets `redis_client = None` and discards the reference, but the `aioredis` client object that was allocated at line 98 is never closed. If the connection was established before the ping exception (e.g., AUTH failure, NOAUTH, wrong database), the underlying socket is abandoned without calling `aclose()`. Add `await redis_client.aclose()` in the except block before nulling the reference.

### SHOULD FIX

- [`backend/app/services/cache_invalidation.py:267-275`] `get_cache_publisher()` allocates a fresh `CacheInvalidationPublisher()` instance every time it is called before `init_cache_publisher` runs. Any caller that caches the returned instance before lifespan completes will hold a no-op publisher forever, even after the real one is initialised. A single module-level no-op sentinel (`_NOOP_PUBLISHER = CacheInvalidationPublisher()`) returned from the fallback branch would be safer and eliminates repeated allocation.

- [`backend/app/services/role.py:489`] `update_role` publishes without `tenant_id` — the call has no `tenant_id` kwarg so it defaults to `None`. `create_role` (line 477) and `assign_role_to_user` (line 499) correctly pass `tenant_id`. Roles can be tenant-scoped; consumers filtering on `tenant_id` will miss cache invalidation for tenant-scoped role updates.

- [`backend/app/services/user.py:653-656`] Remove-user-from-tenant publishes `operation="update"` — removing a tenant membership is a removal/unassignment, and the documented schema lists `"unassign"` as a valid operation. `RoleService.unassign_role_from_user` correctly uses `"unassign"`. Using `"update"` here creates inconsistency; consumers cannot distinguish membership removal from a field edit.

- [`backend/app/main.py:115-120`] `shutdown_cache_publisher()` and `redis_client.aclose()` share a single `try/except` block. Although `shutdown_cache_publisher()` currently cannot raise (it is a `global _publisher = None` assignment), if it ever does, `redis_client.aclose()` will be skipped because the except clause absorbs the exception and continues. The two operations should be in separate try/finally blocks to guarantee both execute.

- [`backend/app/services/cache_invalidation.py:230-243`] `publish_batch` constructs the event dict inline and duplicates the `datetime.now(timezone.utc).isoformat()` call that exists in `_build_event`. The batch event diverges from the standard schema independently. If the timestamp format or timezone handling is ever changed in `_build_event`, the batch path will silently produce inconsistent timestamps. Extract a shared `_now_iso()` helper or use `_build_event` for the common fields.

- [`backend/tests/unit/test_cache_invalidation.py:802-819`] `TestSingletonLifecycle` tests mutate module-level global state without using fixtures to guarantee cleanup. If `test_init_and_get_returns_same_instance` fails before reaching its `shutdown_cache_publisher()` cleanup call, all subsequent tests in the file inherit a dirty singleton. Use `setup_method`/`teardown_method` or a pytest fixture with `autouse=True` to guarantee the singleton is reset regardless of test outcome.

- [`backend/app/services/cache_invalidation.py:280-282`] `init_cache_publisher` performs a bare `global _publisher` assignment with no guard against double-initialisation. If called twice (e.g., in tests that forget to call `shutdown_cache_publisher` between runs), the first publisher's Redis client reference is silently replaced without being closed. The old client leaks. Add a warning log or raise if `_publisher` is already set to a non-None value.

### NITPICK

- [`backend/app/main.py:74`] `import os` added solely for `os.getenv("REDIS_URL")`. The project presumably has a settings/config object that validates env vars at startup. Using raw `os.getenv` bypasses any such validation layer and is inconsistent with how the rest of the app likely reads configuration.

- [`backend/app/services/cache_invalidation.py:203`] Constructor parameter is named `redis_client` but the stored attribute is `self._redis` — the shortened name is slightly inconsistent and marginally harder to search for.

- [`backend/app/services/cache_invalidation.py:192`] `CHANNEL = "identity:changes"` is exported implicitly. If downstream consumers import this constant, they are coupling to the internal module path. Worth noting in the module docstring that `CHANNEL` is part of the stable contract.

- [`backend/app/services/inbound_sync.py:305`] `publisher: CacheInvalidationPublisher | None = None` default is consistent with all other services in this PR, but the `if self._publisher:` guard scattered through business logic is noise. A null-object pattern (always inject a real or no-op publisher, never `None`) would eliminate all seven guard clauses across the six service files.

**Comment by jamescrowley321** [comment]
## Review: Edge Case Hunter

### Findings

| Location | Trigger Condition | Guard Snippet | Consequence |
|----------|-------------------|---------------|-------------|
| `backend/app/services/inbound_sync.py:306` | `_handle_user_deleted` webhook deactivates the user but publishes `operation="sync"` instead of `"deactivate"`; the documented schema lists `"deactivate"` as a distinct operation | Change to `operation="deactivate"` | [WRONG] Downstream cache consumers dispatching on operation type see a sync event, not a deactivation; deactivated users may not be invalidated correctly |
| `backend/app/main.py:87-89` | `shutdown_cache_publisher()` and `redis_client.aclose()` share one `try` block; if `shutdown_cache_publisher()` raises before `aclose()` is reached, the Redis connection is never closed | Split into two sequential `try/finally` blocks so `aclose()` always runs | [DEGRADED] Redis connection leaks on shutdown if `shutdown_cache_publisher` raises (currently cannot raise, but is a structural fragility) |
| `backend/app/services/cache_invalidation.py:114-122` | `get_cache_publisher()` called before `init_cache_publisher()` (e.g., after `shutdown_cache_publisher()` sets `_publisher=None` mid-flight) returns a fresh throwaway no-op instance per call rather than the singleton | Assign the fallback to `_publisher`, or document that only the lifespan path is supported | [DEGRADED] Each pre-init call returns a distinct throwaway object; services injected at DI time hold a different no-op instance than any future call; behaviorally safe but identity is inconsistent and can mask misconfiguration in tests |

### Summary
- Unhandled paths found: 3
- Critical (crash/data loss): 0
- Non-critical (wrong result/degraded): 3

**Comment by jamescrowley321** [comment]
## Review: Acceptance Auditor

### PASS

- [AC-3.3.2] Redis unavailability — all exceptions in `publish()` and `publish_batch()` are caught with a bare `except Exception`, logged as warnings, and never re-raised. The `repo.commit()` call precedes every `publish()` call in every service, so the Postgres write is always committed before Redis is attempted. A `None` redis client (set when `REDIS_URL` is absent or Redis connection fails at startup) causes `publish()` to return immediately without any network attempt. App startup at `backend/app/main.py:95-106` gracefully handles connection failure by setting `redis_client=None` before calling `init_cache_publisher`. Canonical writes are structurally unaffected.
  - Implemented at `backend/app/services/cache_invalidation.py:62-68` (None guard + except), `backend/app/main.py:92-106` (startup degradation)
  - Tested at `backend/tests/unit/test_cache_invalidation.py:56` (no-op when client is None), `backend/tests/unit/test_cache_invalidation.py:62` (exception swallowed, warning logged, no raise), `backend/tests/unit/test_cache_invalidation.py:111` (batch swallows exception)

- [AC-3.3.3] Subscriber documentation — the module-level docstring at `backend/app/services/cache_invalidation.py:1-24` documents: the channel name (`identity:changes`), the full JSON event schema with allowed values for each field, and key patterns subscribers should use to identify affected records and scope invalidation to a tenant. The schema is validated by tests that deserialize actual published payloads and assert all fields are present with correct values.
  - Implemented at `backend/app/services/cache_invalidation.py:9-23`
  - Tested at `backend/tests/unit/test_cache_invalidation.py:32-54` (channel name, all schema fields including timestamp and null tenant_id verified against a live publish call)

### FAIL

- None

### PARTIAL

- [AC-3.3.1] Publish on write — `CacheInvalidationPublisher.publish()` is called after `repo.commit()` in all four canonical entity services (user, role, permission, tenant) and for all write operations. The event includes all required fields (entity_type, entity_id, operation, tenant_id, timestamp). Channel is `identity:changes`.
  - Implemented at: `backend/app/services/user.py:83,153,200,243,284`; `backend/app/services/role.py:101,168,232,299,332,368`; `backend/app/services/permission.py:74,140,174`; `backend/app/services/tenant.py:73`
  - **What's done:** Implementation is present and correct in all four services. `test_user_service.py:TestCacheInvalidationPublishing` (lines 478-532) covers UserService create + deactivate publish assertions, and a guard test confirms no publish on failure. `test_cache_invalidation.py:TestPublish` validates the full event schema for the `publish()` method itself.
  - **What's missing:** `backend/tests/unit/test_role_service.py`, `backend/tests/unit/test_permission_service.py`, and `backend/tests/unit/test_tenant_service.py` contain zero references to `CacheInvalidationPublisher` or `publisher`. There are no tests that would fail if the `if self._publisher: await self._publisher.publish(...)` blocks were removed from `role.py`, `permission.py`, or `tenant.py`. Three of four canonical entity types lack service-level publish tests.

### SCOPE CREEP

- `backend/app/services/inbound_sync.py:318-319`, `:329`, `:339`, `:349` — `InboundSyncService` publishes cache events for flow-sync and webhook-triggered user mutations (operation `"sync"`). AC-3.3.1 specifies "canonical write operations (create, update, delete on user/role/permission/tenant)." Inbound sync is a secondary write path not listed in the AC's entity/operation enumeration. The publisher is wired via `backend/app/dependencies/identity.py:56`. Not traceable to any AC.

- `backend/app/services/reconciliation.py:156-157` — `ReconciliationService.run()` calls `publish_batch()` after a reconciliation pass with changes. Reconciliation is not a canonical write operation per the spec. Additionally, `publish_batch()` emits events with `entity_type="batch"` and `entity_id="reconciliation"`, which fall outside the documented schema in AC-3.3.3 (entity_type is specified as `user | role | permission | tenant`). Wired via `backend/app/dependencies/identity.py:57`. Not traceable to any AC.

### Architecture Violations

- None identified. `CacheInvalidationPublisher` is injected into domain services via constructor, consistent with D4 (constructor injection, onion architecture). The publisher is always called after `await self._repository.commit()`, preserving the D7 write-through Postgres-first semantic. No publisher calls occur in the repository layer (inner) or adapter layer (outer). The singleton pattern mirrors the existing `descope.py` module pattern. The Redis import is gated behind `TYPE_CHECKING` to avoid a hard load-time import.

### Summary

- Total ACs: 3
- Pass: 2 | Fail: 0 | Partial: 1

The partial is AC-3.3.1: implementation is complete and correct across all four canonical entity services, but three of the four service test files (role, permission, tenant) have no publisher assertions. Removing the publish calls from those three services would not be caught by the test suite.

**Comment by jamescrowley321** [comment]
## Review: Security (Sentinel)

### BLOCK (must fix before merge)

- none

### WARN (should fix)

**From prior review (Story 3.1 — still open):**

- **W1** [`backend/app/routers/internal.py:50`] `WebhookPayload.data` is an unvalidated `dict` with no schema enforcement. Webhook handlers extract fields via `.get()` with empty-string defaults. A malformed payload silently returns `Ok({"status": "skipped"})` rather than an error, making production debugging difficult. If Descope changes its payload shape, the endpoint silently stops processing events with no alerting. Consider adding per-event-type Pydantic models or at minimum structured warning logs when required fields are absent for known event types.

- **W2** [`backend/app/routers/internal.py:36-50`] No input length constraints on request model fields. `user_id` is a bare `str` with no `max_length`. `name`, `given_name`, `family_name` are similarly unconstrained. `WebhookPayload.data` is an unbounded `dict`. An attacker with the shared secret could submit arbitrarily large payloads. Add `max_length` constraints on Pydantic fields and consider a body size limit.

- **W3** [`backend/app/middleware/factory.py:79-80`] Gateway mode has no layered protection for internal endpoints. Until network-level isolation is implemented, internal endpoints in gateway mode rely solely on shared secrets with no additional defense. This should be documented as a deployment prerequisite.

- **W4** [`backend/app/services/inbound_sync.py:110-112,151-153`] Commit failures return `Conflict` error type regardless of cause. Bare `except Exception` blocks around `commit()` map all exceptions (including network errors, connection timeouts) to `Conflict` responses. A database connectivity failure should not be presented as a conflict.

- **W5** [`backend/app/services/inbound_sync.py:223-225`] The `_handle_user_updated` webhook path reads `email` from the untyped webhook `data` dict and writes it directly to the user model without email format validation (unlike the flow sync path which uses Pydantic `EmailStr`). A malformed email value would be persisted to the database.

**New findings — Story 3.3 (cache invalidation):**

- **W6** [`backend/app/main.py:93-98`] Redis URL is consumed without scheme validation. `aioredis.from_url(redis_url)` accepts any URL scheme the `redis` library supports, including `unix://`. If `REDIS_URL` is misconfigured or injected (e.g., via a compromised secrets manager), the application will silently attempt to connect to a non-Redis endpoint. The graceful-degradation catch block swallows the error, which is intentional, but a badly formed URL pointing at an internal metadata endpoint (e.g., `http://169.254.169.254/...`) would cause a connection attempt to that host at startup without any scheme allowlist.
  Mitigation: Validate `redis_url` scheme before calling `aioredis.from_url()` — e.g., assert scheme in `{"redis", "rediss"}`.

- **W7** [`backend/app/services/cache_invalidation.py:267-275`] `get_cache_publisher()` has a singleton gap: if called before `init_cache_publisher()` (e.g., in tests that import the module without going through app lifespan), it returns a throwaway no-op `CacheInvalidationPublisher()` that is not stored in `_publisher`. A consumer that caches this returned reference before init completes will hold a permanently no-op publisher and silently miss all cache invalidation events after Redis connects. The security implication is that downstream caches may serve stale identity or permission data indefinitely with no indication of the misconfiguration.
  Mitigation: Document init ordering requirement; add a warning log if `get_cache_publisher()` is called before `init_cache_publisher()`.

### INFO (acceptable risk)

- **I1** PII handling in OTel traces is properly mitigated. `inbound_sync.py` uses `_hash_email(email)` (SHA-256 truncated) for span attributes rather than raw email. UUIDs in span attributes are not PII.

- **I2** PII in log messages is properly handled. Logger calls include only `user.id` (a UUID). Skip-reason logs emit `list(data.keys())` rather than raw data values, avoiding PII leakage from webhook payloads.

- **I3** HMAC implementation is cryptographically correct. `verify_hmac_signature` uses `hmac.new()` with SHA-256 and `hmac.compare_digest()` for timing-safe comparison. Secret read once at import time.

- **I4** Flow sync shared secret implementation is correct. `verify_flow_sync_secret` uses `hmac.compare_digest()` for timing-safe string comparison. Empty/missing secret configuration is rejected with 401.

- **I5** Rate limiting is applied to both internal endpoints (default: 10/minute via `RATE_LIMIT_AUTH`). SlowAPI middleware is active in standalone mode.

- **I6** `excluded_prefixes` implementation is not vulnerable to path traversal. Uses `str.startswith(tuple)`; ASGI servers normalize paths before middleware.

- **I7** Fail-closed behavior on missing secrets. When `DESCOPE_WEBHOOK_SECRET` or `DESCOPE_FLOW_SYNC_SECRET` is empty/unset, the respective endpoints reject all requests with 401.

- **I8** No SQL injection risk. All data flows through SQLAlchemy ORM parameterized queries.

- **I9** Secrets loaded at import time (`internal.py:29-30`) are not reloadable without restart. Acceptable for current deployment model.

- **I10** Cache invalidation events contain no credentials or PII. Published events carry only UUIDs and enum-like strings. The `stats` dict in `publish_batch` contains only server-side integer counters — no user-supplied data flows into the Redis channel payload. A compromised Redis subscriber learns only that a mutation occurred on a given entity type/ID, not the content of the change.

- **I11** Redis auth is configured in docker-compose. `REDIS_URL` includes password authentication; Redis port 6379 is bound to `127.0.0.1` only, limiting exposure to the Docker network. The default `changeme` password is documented as overridable via `REDIS_PASSWORD`.

- **I12** Cache invalidation failure is genuinely fire-and-forget. All `publish()` and `publish_batch()` calls are wrapped in `try/except Exception`. Failures never propagate to the service layer and never roll back committed DB transactions. A Redis outage cannot cause data inconsistency in the canonical store.

- **I13** No SSRF risk from Redis URL under normal operation. The URL is taken from a server-side environment variable, not from any user-supplied request parameter. (W6 above covers the env-injection edge case.)

- **I14** `publish_batch` stats integrity. The `stats` dict is initialized with hardcoded string keys and integer zero values; only `+= 1` increments modify it before publication. No user-controlled data reaches the published stats payload.

- **I15** `shutdown_cache_publisher()` correctly nulls the singleton before Redis client is closed (`main.py:87-89`). Subsequent calls to `get_cache_publisher()` after shutdown return a new no-op instance, preventing use-after-close of the Redis client.

### Summary
- BLOCK: 0 | WARN: 7 | INFO: 15
- Overall: PASS

**Comment by jamescrowley321** [comment]
## Review: Red Team (Viper)

### Attack Surface

1. **POST /api/internal/users/sync** -- New endpoint for Descope Flow HTTP Connector user synchronisation. Accepts `user_id`, `email`, `name`, `given_name`, `family_name`. Protected by shared secret (`X-Flow-Secret` header validated against `DESCOPE_FLOW_SYNC_SECRET` via `verify_flow_sync_secret` dependency at `internal.py:98`). JWT auth bypassed via `excluded_prefixes`. Rate-limited at `RATE_LIMIT_AUTH` (default 10/minute).
2. **POST /api/internal/webhooks/descope** -- New endpoint for Descope audit webhook events. Accepts `event_type` and arbitrary `data` dict. Protected by HMAC-SHA256 (`X-Descope-Webhook-S256` header validated against `DESCOPE_WEBHOOK_SECRET`). JWT auth bypassed. Rate-limited at `RATE_LIMIT_AUTH`.
3. **JWT bypass via excluded_prefixes** -- Entire `/api/internal/` prefix excluded from `TokenValidationMiddleware` in `factory.py:74-76`. Any future endpoint registered under this prefix automatically inherits the bypass.
4. **Shared secret authentication on flow sync** -- `verify_flow_sync_secret` dependency at `internal.py:56-69` performs timing-safe comparison of `X-Flow-Secret` header against `_FLOW_SYNC_SECRET` module-level variable.
5. **HMAC-SHA256 authentication on webhook** -- `verify_hmac_signature` dependency at `internal.py:75-92` computes HMAC over raw request body and compares against `X-Descope-Webhook-S256` header.
6. **InboundSyncService** -- New service layer at `inbound_sync.py` that creates/updates User records and IdPLink records in Postgres based on external input.
7. **IdPLinkRepository** -- New repository at `idp_link.py` for IdP link CRUD, with unique constraints on `(provider_id, external_sub)` and `(user_id, provider_id)`.
8. **ProviderRepository** -- New repository at `provider.py` for provider lookup by type/name.
9. **Startup secret validation** -- `_warn_missing_secrets()` in `main.py:35-42` logs warnings at startup when `DESCOPE_WEBHOOK_SECRET` or `DESCOPE_FLOW_SYNC_SECRET` are not set. Both endpoints fail closed (401) when secrets are missing.

### Findings

#### MEDIUM -- Unbounded string length on user_id and name fields enables storage abuse

- **Location**: `backend/app/routers/internal.py:39-43` (`FlowSyncRequest` model), `backend/app/models/identity/user.py:69` (`external_sub` column), `backend/app/models/identity/user.py:30-31` (`given_name`/`family_name` columns)
- **Attack scenario**:
  1. Attacker obtains the `DESCOPE_FLOW_SYNC_SECRET` shared secret (e.g., leaked in logs, config repo, or insider threat).
  2. Attacker sends POST to `/api/internal/users/sync` with `user_id` set to a multi-megabyte string and valid `X-Flow-Secret` header.
  3. `FlowSyncRequest.user_id` is typed as `str` with no `max_length` constraint. The value flows through to `IdPLink.external_sub` which is `sa.Column(sa.String, nullable=False)` -- unbounded VARCHAR.
  4. Similarly, `name`, `given_name`, `family_name` have no length limits and are stored in unbounded `sa.String` columns.
  5. Repeated requests (up to 10/minute per IP due to rate limiting) with large payloads bloat the database.
- **Prerequisites**: Knowledge of `DESCOPE_FLOW_SYNC_SECRET`. Rate limiting (10/min) constrains throughput.
- **CVSS v3.1**: 4.3 (AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:L)
- **Remediation**: Add `max_length` constraints to `FlowSyncRequest` fields: `user_id: str = Field(max_length=255)`, `name: str | None = Field(default=None, max_length=512)`, `given_name: str | None = Field(default=None, max_length=255)`, `family_name: str | None = Field(default=None, max_length=255)`. Also add `String(255)` or similar length to the SQLAlchemy column definitions for `external_sub`, `given_name`, `family_name` in the ORM models.

#### MEDIUM -- WebhookPayload.data typed as unvalidated dict accepts arbitrary nested structures

- **Location**: `backend/app/routers/internal.py:49-50` (`WebhookPayload` model), `backend/app/services/inbound_sync.py:186-289` (webhook handlers)
- **Attack scenario**:
  1. Attacker compromises the HMAC webhook secret.
  2. The `data` field accepts any JSON object with no schema validation per event type.
  3. In `_handle_user_updated` (line 223-236), `data.get("email")` could return a non-string type (e.g., a list or dict). This value is assigned directly to `user.email` at line 225. Depending on SQLAlchemy's type coercion for the `sa.String` column, this could raise an unhandled exception (500 error) or store a stringified representation.
  4. `data.get("name")` could return a non-string. When passed to `name.split(" ", 1)` at line 234, a non-string raises `AttributeError`, which propagates as an unhandled 500.
  5. A deeply nested or very large `data` dict could consume excessive memory during JSON parsing.
- **Prerequisites**: Valid HMAC signature (requires `DESCOPE_WEBHOOK_SECRET`).
- **CVSS v3.1**: 4.3 (AV:N/AC:L/PR:L/UI:N/S:U/C:N/I:N/A:L)
- **Remediation**: Define typed inner models for the `data` field per event type, or validate field types in the service layer before use. At minimum, add type guards: `if isinstance(email, str)` before assignment, `if isinstance(name, str)` before calling `.split()`.

#### LOW -- Module-level secret caching prevents runtime secret rotation

- **Location**: `backend/app/routers/internal.py:29-30`
- **Attack scenario**:
  1. A secret is compromised and the operator rotates `DESCOPE_FLOW_SYNC_SECRET` or `DESCOPE_WEBHOOK_SECRET` in the environment.
  2. The application continues using the old (compromised) secret because `_FLOW_SYNC_SECRET` and `_WEBHOOK_SECRET` are read once at module import time.
  3. The operator must restart the application process for the new secret to take effect.
  4. During the window between rotation and restart, the compromised secret remains valid.
- **Prerequisites**: Secret compromise plus operational delay in restarting. This is a deployment hygiene issue rather than a direct exploit vector.
- **CVSS v3.1**: 2.6 (AV:N/AC:H/PR:H/UI:N/S:U/C:N/I:L/A:N)
- **Remediation**: Acceptable trade-off. The comment at line 28 ("Read secrets once at import time; validated at startup in lifespan") documents the intentional design. If runtime rotation is later desired, read from environment per-request with TTL caching, or integrate a secrets manager.

#### LOW -- Webhook log messages include data.keys() which could contain attacker-controlled strings

- **Location**: `backend/app/services/inbound_sync.py:192,207,260`
- **Attack scenario**:
  1. Attacker sends a webhook with crafted data keys containing malicious strings (e.g., ANSI escape sequences, HTML/JS payloads).
  2. `logger.warning("... keys=%s", list(data.keys()))` writes attacker-controlled key names to application logs.
  3. If logs are rendered in a web-based log viewer without sanitization, the key names could be exploited for log injection.
- **Prerequisites**: Valid HMAC signature plus a vulnerable log viewer rendering the log entries.
- **CVSS v3.1**: 2.0 (AV:N/AC:H/PR:L/UI:R/S:U/C:N/I:L/A:N)
- **Remediation**: Minor concern. The approach of logging only keys (not values) is already good practice. For additional hardening, log only the count of unexpected keys or sanitize key names.

### Non-findings (verified secure)

- **Flow sync authentication**: The flow sync endpoint IS authenticated. `internal.py:98` specifies `dependencies=[Depends(verify_flow_sync_secret)]`, which validates the `X-Flow-Secret` header against `_FLOW_SYNC_SECRET` using `hmac.compare_digest()` (line 68). Missing or incorrect secret returns 401. Tests at `test_internal_router.py:114-146` cover invalid secret, missing header, and unconfigured secret scenarios. PASS.
- **Timing-safe comparison**: Both `verify_flow_sync_secret` (line 68) and `verify_hmac_signature` (line 91) use `hmac.compare_digest()`, preventing timing side-channel attacks on secret/signature comparison. PASS.
- **HMAC construction**: `hmac.new(secret.encode(), body, hashlib.sha256).hexdigest()` at line 89 correctly computes HMAC-SHA256 over the raw request body with the secret as the HMAC key. PASS.
- **Fail-closed on missing secrets**: Both `verify_flow_sync_secret` (lines 64-66) and `verify_hmac_signature` (lines 84-86) reject requests with 401 when the respective secret is empty or unset. Startup warnings in `main.py:39-42` alert operators. PASS.
- **Email PII in OTel spans**: The service uses `_hash_email()` at `inbound_sync.py:27-29` to SHA-256 hash emails before setting span attributes (line 66: `span.set_attribute("user.email_hash", _hash_email(email))`). No raw email addresses in traces. PASS.
- **SQL injection**: All queries use parameterized SQLAlchemy expressions (e.g., `IdPLink.external_sub == external_sub`), never string interpolation. PASS.
- **IDOR on user records**: Users are looked up via IdP link (`provider_id + external_sub`) or by email, never by attacker-supplied canonical UUIDs. User IDs in responses are server-generated. PASS.
- **Rate limiting**: Both endpoints decorated with `@limiter.limit(RATE_LIMIT_AUTH)` (default 10/minute) at `internal.py:99` and `internal.py:123`. The rate limiter falls back to IP address when no JWT claims are present (as expected for internal endpoints). PASS.
- **IdPLink unique constraints**: `uq_idp_links_provider_external_sub` prevents duplicate IdP links for the same provider/subject. `uq_idp_links_user_provider` prevents a user from having multiple links to the same provider. Database-level enforcement. PASS.
- **Repository transaction ownership**: `IdPLinkRepository.create()` at `idp_link.py:42-47` does NOT call `rollback()` on `IntegrityError` -- it raises `RepositoryConflictError` and lets the service layer control the transaction. Comment at line 41 explicitly documents this. PASS.
- **Commit error handling**: The service wraps all `commit()` calls in try/except blocks (e.g., lines 108-112, 149-153 of `inbound_sync.py`), returning typed `Result` errors instead of propagating raw exceptions as 500s. PASS.

### Summary

- Attack surface elements: 9
- Findings: 0 critical, 0 high, 2 medium, 2 low
- Overall: PASS

The implementation has a solid security posture. Both internal endpoints are authenticated (shared secret for flow sync, HMAC-SHA256 for webhooks), use timing-safe comparisons, fail closed when secrets are unconfigured, and are rate-limited. Email PII is hashed before inclusion in OTel spans. Repository transaction ownership is correctly managed. The two medium findings (unbounded string lengths and unvalidated webhook data types) are defense-in-depth improvements that require a pre-existing secret compromise to exploit. No critical or high-severity vulnerabilities were identified.

File: backend/app/services/cache_invalidation.py [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        tenant_id: UUID | str | None = None,
+    ) -> None:
+        """Publish a single change event. Silent on failure (AC-3.3.2)."""
+        if self._redis is None:
+            return
+        event = self._build_event(entity_type, entity_id, operation, tenant_id)
+        try:
+            await self._redis.publish(CHANNEL, json.dumps(event))
+        except Exception:
+            logger.warning("Redis publish failed for %s/%s", entity_type, entity_id, exc_info=True)
```
Comment (jamescrowley321) [inline backend/app/services/cache_invalidation.py]: Make this an error.

File: backend/app/services/cache_invalidation.py [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            "operation": operation,
+            "tenant_id": str(tenant_id) if tenant_id is not None else None,
+            "timestamp": datetime.now(timezone.utc).isoformat(),
+        }
+
+
+# ---------------------------------------------------------------------------
+# Module-level singleton (mirrors descope.py pattern)
+# ---------------------------------------------------------------------------
+_publisher: CacheInvalidationPublisher | None = None
```
Comment (jamescrowley321) [inline backend/app/services/cache_invalidation.py:105]: This pattern is a code smell.

File: backend/app/services/cache_invalidation.py (hunk 1 above)
Comment (jamescrowley321) [reply backend/app/services/cache_invalidation.py]: Escalated to `logger.error` in e9598ed — AC-3.3.2 "never raised" contract preserved, `publish_batch` got the same treatment, docstrings and unit test assertions updated to match.

File: backend/app/services/cache_invalidation.py (hunk 2 above)
Comment (jamescrowley321) [reply backend/app/services/cache_invalidation.py:105]: Agreed — module-level singletons are code smell. Leaving as-is in this PR because T150 / #216 is a dedicated refactor task that rips out both this publisher singleton **and** the matching `descope.py` singleton and replaces them with `app.state` DI. Doing half the refactor here would either ship an inconsistent pattern or blow up the scope of story 3.3. The canonical loop will pick up #216 as its next task.

**Comment by jamescrowley321** [comment]
Subsumed by #226 (merge commit 791d220), which landed the entire canonical chain (stories 2.2–4.4) on main in a single linearized commit. The story-3.3 Redis pub/sub work is included, plus the cache_invalidation log-level fix (warning → error) that was addressed in commit e9598ed was cherry-picked forward into #226. The `canonical/story-3.3-redis-pubsub` branch is being kept until cleanup is explicitly confirmed.
